plugins/chat/lib/chat/channel_archive_service.rb

```diff
@@ -107,22 +108,130 @@ def execute
+              thread_om = thread.first
+
+              if !thread_om.nil?
+                thread_size = thread.size - 1
+                last_thread_index = 0
+                iterations = (message_chunk.size.to_f / (ARCHIVED_MESSAGES_PER_POST - 1)).ceil
+
+                iterations.times do |index|
+                  if last_thread_index != thread_size
+                    if index == 0
+                      thread_index = thread.index(post_last_message)
+                    else
+                      next_post_last_message =
+                        message_chunk[(ARCHIVED_MESSAGES_PER_POST * (index + 1)) - index]
+                      if next_post_last_message&.thread_id == post_last_message&.thread_id
+                        thread_index = last_thread_index + ARCHIVED_MESSAGES_PER_POST - 1
+                      else
+                        thread_index = thread_size
+                      end
+                    end
+
+                    range =
+                      I18n.t(
+                        "chat.transcript.split_thread_range",
+                        start: last_thread_index + 1,
+                        end: thread_index,
+                        total: thread_size,
+                      )
+
+                    thread_ranges[thread_om.thread_id] ||= []
+                    thread_ranges[thread_om.thread_id] << range
+                    last_thread_index = thread_index
+                  end
+                end
+              end
+            end
+
+            batch = []
+            batch_thread_added = false
+
+            message_chunk.each do |message|
+              # We duplicate the original message when it spans across multiple posts
+
+              if thread_om.present?
+                if batch.empty? && message_chunk.size > ARCHIVED_MESSAGES_PER_POST &&
+                     message&.thread_id == thread_om&.thread_id && message != thread_om
+                  batch << thread_om
+
+                  batch_thread_ranges[thread_om.id] = thread_ranges[message.thread_id].first
+                  thread_ranges[message.thread_id].slice!(0)
+                elsif thread_ranges.has_key?(message.thread_id) &&
+                      thread_ranges[message.thread_id].present? && batch_thread_added == false
+                  batch_thread_ranges[thread_om.id] = thread_ranges[message.thread_id].first
+                  thread_ranges[message.thread_id].slice!(0)
+
+                  batch_thread_added = true
+                end
+              end
+              if message == thread_om && batch.size + 1 >= ARCHIVED_MESSAGES_PER_POST
+                batch_size = batch.size + 1
+              else
+                batch << message
+                batch_size = batch.size
+              end
+
+              if batch_size >= ARCHIVED_MESSAGES_PER_POST
+                create_post_from_batch(batch, batch_thread_ranges)
+                batch.clear
+              end
```

Comment:
Same we should extract this in one function with a descriptive name

Reply:
I gave it a couple of tries because I wanted to do that, but I couldn't find anything that made much sense to me. Since the code requires a global context from other iterations in loops and such, it's hard to isolate things. Otherwise, you end up passing a ton of attributes to a function only to return them back, which I don't feel is any clearer